ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2.hs

```diff
@@ -0,0 +1,173 @@
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound.V2
+  ( -- * ObjectDiffusion Inbound client
+    objectDiffusionInbound
+
+    -- * PeerStateAPI
+  , withPeer
+  , PeerStateAPI
+
+    -- * Supporting types
+  , module V2
+  , PeerDecisionChannelsVar
+  , newPeerDecisionChannelsVar
+  , DecisionPolicy (..)
+  ) where
+
+import Control.Concurrent.Class.MonadSTM (MonadSTM, atomically)
+import Control.Monad.Class.MonadThrow
+import Control.Tracer (Tracer, traceWith)
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Set qualified as Set
+import Network.TypedProtocol
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound.V2.Registry
+import Ouroboros.Consensus.MiniProtocol.ObjectDiffusion.Inbound.V2.Types as V2
+import Ouroboros.Network.ControlMessage (ControlMessage (..), ControlMessageSTM)
+import Ouroboros.Network.Protocol.ObjectDiffusion.Inbound
+
+-- TODO: Add checks and validation
+
+-- | A object-diffusion inbound side (client).
+--
+-- The steps are as follow
+-- 1. Block on next decision from the decision logic
+-- 2. Handle any available reply (`goCollect`)
+-- 3. Request new objects if possible (`goReqObjects`)
+-- 4. Request new ids (also responsible for ack) (`goReqIds`)
+-- 5. signal psaOnDecisionCompleted (as part of `goReqIds{Blocking,NonBlocking}`)
+-- And loop again
+-- We need to make sure we don't go again into `goIdle` until `psaOnDecisionCompleted` has been called
+objectDiffusionInbound ::
+  forall objectId object m.
+  ( MonadThrow m
+  , MonadSTM m
+  ) =>
+  Tracer m (TraceObjectDiffusionInbound objectId object) ->
+  ControlMessageSTM m ->
+  PeerStateAPI m objectId object ->
+  ObjectDiffusionInboundPipelined objectId object m ()
+objectDiffusionInbound
+  tracer
+  controlMessageSTM
+  PeerStateAPI
+    { psaReadDecision
+    , psaOnDecisionCompleted
+    , psaOnRequestIds
+    , psaOnRequestObjects
+    , psaOnReceiveIds
+    , psaOnReceiveObjects
+    } =
+    ObjectDiffusionInboundPipelined $ goIdle Zero
+   where
+    goIdle :: forall (n :: N). Nat n -> InboundStIdle n objectId object m ()
+    goIdle n = WithEffect $ do
+      ctrlMsg <- atomically controlMessageSTM
+      traceWith tracer $ TraceObjectDiffusionInboundReceivedControlMessage ctrlMsg
+      case ctrlMsg of
+        -- The peer selection governor is asking us to terminate the connection.
+        Terminate ->
+          pure $ terminateAfterDrain n
```

Comment:
There seem to be no tracing on this path to signal termination. Is this expected?

Reply:
IIRC, I've been said (by Alex?) that we only need to improve tracing if the need shows up.  
But maybe are they some more precise guidelines for tracing somewhere?